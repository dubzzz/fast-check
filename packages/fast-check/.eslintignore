bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs