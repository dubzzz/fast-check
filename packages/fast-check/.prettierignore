bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/