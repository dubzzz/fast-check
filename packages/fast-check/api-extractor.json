{
  "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
  "compiler": {
    "tsconfigFilePath": "<projectFolder>/tsconfig.publish.types.json"
  },
  "tsdocMetadata": {
    "enabled": false
  },
  "mainEntryPointFilePath": "lib/types/fast-check-default.d.ts",
  "dtsRollup": {
    "enabled": false
  },
  "docModel": {
    "enabled": true,
    "apiJsonFilePath": "docs/fast-check.api.json"
  },
  "apiReport": {
    "enabled": false
  },
  "messages": {
    "compilerMessageReporting": {
      "default": {
        "logLevel": "error"
      }
    },
    "extractorMessageReporting": {
      "default": {
        "logLevel": "error"
      },
      "ae-unresolved-link": {
        "logLevel": "warning"
      }
    },
    "tsdocMessageReporting": {
      "default": {
        "logLevel": "error"
      }
    }
  }
}
