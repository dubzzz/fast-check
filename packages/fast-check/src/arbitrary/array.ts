import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { ArrayArbitrary } from './_internals/ArrayArbitrary';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import {
  MaxLengthUpperBound,
  maxGeneratedLengthFromSizeForArbitrary,
} from './_internals/helpers/MaxLengthFromMinLength';
import type { DepthIdentifier } from './_internals/helpers/DepthContext';

/**
 * Constraints to be applied on {@link array}
 * @remarks Since 2.4.0
 * @public
 */
export interface ArrayConstraints {
  /**
   * Lower bound of the generated array size
   * @defaultValue 0
   * @remarks Since 2.4.0
   */
  minLength?: number;
  /**
   * Upper bound of the generated array size
   * @defaultValue 0x7fffffff â€” _defaulting seen as "max non specified" when `defaultSizeToMaxWhenMaxSpecified=true`_
   * @remarks Since 2.4.0
   */
  maxLength?: number;
  /**
   * Define how large the generated values should be (at max)
   *
   * When used in conjonction with `maxLength`, `size` will be used to define
   * the upper bound of the generated array size while `maxLength` will be used
   * to define and document the general maximal length allowed for this case.
   *
   * @remarks Since 2.22.0
   */
  size?: SizeForArbitrary;
  /**
   * When receiving a depth identifier, the arbitrary will impact the depth
   * attached to it to avoid going too deep if it already generated lots of items.
   *
   * In other words, if the number of generated values within the collection is large
   * then the generated items will tend to be less deep to avoid creating structures a lot
   * larger than expected.
   *
   * For the moment, the depth is not taken into account to compute the number of items to
   * define for a precise generate call of the array. Just applied onto eligible items.
   *
   * @remarks Since 2.25.0
   */
  depthIdentifier?: DepthIdentifier | string;
}

/**
 * Extra but internal constraints that can be passed to array.
 * This extra set is made of constraints mostly targets experimental and internal features not yet mature to be exposed.
 * @internal
 */
export interface ArrayConstraintsInternal<T> extends ArrayConstraints {
  /**
   * Extra user-definable and hardcoded values.
   * Each entry in the array could be used to build the final generated value outputed by the arbitrary of array on generate.
   * Each entry must have at least one element of type T into it.
   * Each T must be a value acceptable for the arbitrary passed to the array.
   */
  experimentalCustomSlices?: T[][];
}

/**
 * For arrays of values coming from `arb`
 *
 * @param arb - Arbitrary used to generate the values inside the array
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function array<T>(arb: Arbitrary<T>, constraints: ArrayConstraints = {}): Arbitrary<T[]> {
  const size = constraints.size;
  const minLength = constraints.minLength || 0;
  const maxLengthOrUnset = constraints.maxLength;
  const depthIdentifier = constraints.depthIdentifier;
  const maxLength = maxLengthOrUnset !== undefined ? maxLengthOrUnset : MaxLengthUpperBound;
  const specifiedMaxLength = maxLengthOrUnset !== undefined;
  const maxGeneratedLength = maxGeneratedLengthFromSizeForArbitrary(size, minLength, maxLength, specifiedMaxLength);
  const customSlices = (constraints as ArrayConstraintsInternal<T>).experimentalCustomSlices || [];
  return new ArrayArbitrary<T>(arb, minLength, maxGeneratedLength, maxLength, depthIdentifier, undefined, customSlices);
}
export { array };
