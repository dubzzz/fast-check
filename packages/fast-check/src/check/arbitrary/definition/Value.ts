import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
