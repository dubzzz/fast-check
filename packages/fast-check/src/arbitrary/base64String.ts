import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import { base64 } from './base64';
import { MaxLengthUpperBound } from './_internals/helpers/MaxLengthFromMinLength';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
import { stringToBase64Mapper, stringToBase64Unmapper } from './_internals/mappers/StringToBase64';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For base64 strings
 *
 * A base64 string will always have a length multiple of 4 (padded with =)
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function base64String(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const { minLength: unscaledMinLength = 0, maxLength: unscaledMaxLength = MaxLengthUpperBound, size } = constraints;

  const minLength = unscaledMinLength + 3 - ((unscaledMinLength + 3) % 4);
  const maxLength = unscaledMaxLength - (unscaledMaxLength % 4);
  const requestedSize = constraints.maxLength === undefined && size === undefined ? '=' : size;

  if (minLength > maxLength) throw new Error('Minimal length should be inferior or equal to maximal length');
  if (minLength % 4 !== 0) throw new Error('Minimal length of base64 strings must be a multiple of 4');
  if (maxLength % 4 !== 0) throw new Error('Maximal length of base64 strings must be a multiple of 4');

  const charArbitrary = base64();
  const experimentalCustomSlices = createSlicesForString(charArbitrary, codePointsToStringUnmapper);
  const enrichedConstraints: ArrayConstraintsInternal<string> = {
    minLength,
    maxLength,
    size: requestedSize,
    experimentalCustomSlices,
  };
  return array(charArbitrary, enrichedConstraints)
    .map(codePointsToStringMapper, codePointsToStringUnmapper)
    .map(stringToBase64Mapper, stringToBase64Unmapper);
}
export { base64String };
