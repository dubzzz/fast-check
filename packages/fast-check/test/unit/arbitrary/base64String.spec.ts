import * as fc from 'fast-check';
import { base64String } from '../../../src/arbitrary/base64String';

import {
  assertProduceValuesShrinkableWithoutContext,
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
} from './__test-helpers__/ArbitraryAssertions';

import * as ArrayMock from '../../../src/arbitrary/array';
import { fakeArbitrary } from './__test-helpers__/ArbitraryHelpers';
import { Value } from '../../../src/check/arbitrary/definition/Value';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';
import { sizeForArbitraryArb } from './__test-helpers__/SizeHelpers';

function beforeEachHook() {
  jest.resetModules();
  jest.restoreAllMocks();
  fc.configureGlobal({ beforeEach: beforeEachHook });
}
beforeEach(beforeEachHook);

describe('base64String', () => {
  it('should accept any constraints accepting at least one length multiple of 4', () =>
    fc.assert(
      fc.property(
        fc.nat({ max: 5 }),
        fc.integer({ min: 3, max: 30 }),
        fc.boolean(),
        fc.boolean(),
        (min, gap, withMin, withMax) => {
          // Arrange
          const constraints = { minLength: withMin ? min : undefined, maxLength: withMax ? min + gap : undefined };

          // Act / Assert
          expect(() => base64String(constraints)).not.toThrowError();
        },
      ),
    ));

  it('should reject any constraints not accepting at least one length multiple of 4', () =>
    fc.assert(
      fc.property(fc.nat({ max: 30 }), fc.nat({ max: 2 }), (min, gap) => {
        // Arrange
        const constraints = { minLength: min, maxLength: min + gap };
        let includesMultipleOf4 = false;
        for (let acceptedLength = constraints.minLength; acceptedLength <= constraints.maxLength; ++acceptedLength) {
          includesMultipleOf4 = includesMultipleOf4 || acceptedLength % 4 === 0;
        }
        fc.pre(!includesMultipleOf4);

        // Act / Assert
        expect(() => base64String(constraints)).toThrowError();
      }),
    ));

  it('should always query for arrays that will produce length fitting the requested range', () =>
    fc.assert(
      fc.property(
        fc.nat({ max: 30 }),
        fc.integer({ min: 3, max: 30 }),
        fc.boolean(),
        fc.boolean(),
        (min, gap, withMin, withMax) => {
          // Arrange
          const constraints = { minLength: withMin ? min : undefined, maxLength: withMax ? min + gap : undefined };
          const array = jest.spyOn(ArrayMock, 'array');
          const { instance: arrayInstance, map } = fakeArbitrary();
          array.mockReturnValue(arrayInstance);
          map.mockReturnValue(arrayInstance); // fake map

          // Act
          base64String(constraints);

          // Assert
          expect(array).toHaveBeenCalledTimes(1);
          const constraintsOnArray = array.mock.calls[0][1]!;
          const rounded4 = (value: number) => {
            switch (value % 4) {
              case 0:
                return value;
              case 1:
                return value;
              case 2:
                return value + 2;
              case 3:
                return value + 1;
            }
          };
          if (constraints.minLength !== undefined) {
            expect(constraintsOnArray.minLength).toBeDefined();
            expect(rounded4(constraintsOnArray.minLength!)).toBeGreaterThanOrEqual(constraints.minLength);
          }
          if (constraints.maxLength !== undefined) {
            expect(constraintsOnArray.maxLength).toBeDefined();
            expect(rounded4(constraintsOnArray.maxLength!)).toBeLessThanOrEqual(constraints.maxLength);
          }
          if (constraintsOnArray.minLength !== undefined && constraintsOnArray.maxLength !== undefined) {
            expect(constraintsOnArray.maxLength).toBeGreaterThanOrEqual(constraintsOnArray.minLength);
          }
        },
      ),
    ));

  it('should always forward constraints on size to the underlying arbitrary when provided', () =>
    fc.assert(
      fc.property(
        fc.nat({ max: 5 }),
        fc.integer({ min: 3, max: 30 }),
        fc.boolean(),
        fc.boolean(),
        sizeForArbitraryArb,
        (min, gap, withMin, withMax, size) => {
          // Arrange
          const constraints = {
            minLength: withMin ? min : undefined,
            maxLength: withMax ? min + gap : undefined,
            size,
          };
          const array = jest.spyOn(ArrayMock, 'array');
          const { instance: arrayInstance, map } = fakeArbitrary();
          array.mockReturnValue(arrayInstance);
          map.mockReturnValue(arrayInstance); // fake map

          // Act
          base64String(constraints);

          // Assert
          expect(array).toHaveBeenCalledWith(expect.anything(), expect.objectContaining({ size }));
        },
      ),
    ));
});

describe('base64String (integration)', () => {
  type Extra = { minLength?: number; maxLength?: number };
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(fc.nat({ max: 30 }), fc.integer({ min: 3, max: 30 }), fc.boolean(), fc.boolean())
    .map(([min, gap, withMin, withMax]) => ({
      minLength: withMin ? min : undefined,
      // Minimal gap=3 to ensure we have at least one possible multiple of 4 between min and max
      maxLength: withMax ? min + gap : undefined,
    }));

  const isCorrect = (value: string, extra: Extra) => {
    if (extra.minLength !== undefined) {
      expect(value.length).toBeGreaterThanOrEqual(extra.minLength);
    }
    if (extra.maxLength !== undefined) {
      expect(value.length).toBeLessThanOrEqual(extra.maxLength);
    }
    const padStart = value.indexOf('=');
    const beforeEqualValue = value.substr(0, padStart === -1 ? value.length : padStart);
    const afterEqualValue = value.substr(padStart === -1 ? value.length : padStart);
    for (const c of beforeEqualValue.split('')) {
      expect('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+/'.split('')).toContainEqual(c);
    }
    expect(['', '=', '==']).toContainEqual(afterEqualValue);
  };

  const base64StringBuilder = (extra: Extra) => base64String(extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(base64StringBuilder, { extraParameters });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(base64StringBuilder, isCorrect, { extraParameters });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(base64StringBuilder, { extraParameters });
  });

  // assertShrinkProducesSameValueWithoutInitialContext is not applicable for base64String has some values will not shrink exactly the same way.
  // For instance: 'abcde' will be mapped to 'abcd', with default shrink it will try to shrink from 'abcde'. With context-less one it will start from 'abcd'.

  it.each`
    source                               | constraints
    ${'0123ABC==' /* invalid base 64 */} | ${{}}
    ${'AB==' /* not large enough */}     | ${{ minLength: 5 }}
    ${'0123AB==' /* too large */}        | ${{ maxLength: 7 }}
  `('should not be able to generate $source with fc.base64String($constraints)', ({ source, constraints }) => {
    // Arrange / Act
    const arb = base64String(constraints);
    const out = arb.canShrinkWithoutContext(source);

    // Assert
    expect(out).toBe(false);
  });

  it.each`
    rawValue
    ${'ABCD'}
    ${'0123AB=='}
    ${'01230123012301230123AB=='}
    ${'ABCD'.repeat(50)}
  `('should be able to shrink $rawValue', ({ rawValue }) => {
    // Arrange
    const arb = base64String();
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
