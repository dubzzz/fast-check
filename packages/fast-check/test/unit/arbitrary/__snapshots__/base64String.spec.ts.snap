// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`base64String (integration) > should be able to shrink '0123AB==' 1`] = `
""0123AB=="
├> ""
├> "3AB="
|  ├> "AB=="
|  |  ├> ""
|  |  |  ├> ""
|  |  |  └> ""
|  |  |     └> ""
|  |  ├> ""
|  |  |  └> ""
|  |  └> "AA=="
|  |     ├> ""
|  |     |  └> ""
|  |     └> ""
|  |        └> ""
|  ├> "AAB="
|  |  ├> "AB=="
|  |  |  ├> ""
|  |  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  |     └> ""
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  └> "AA=="
|  |  |     ├> ""
|  |  |     |  └> ""
|  |  |     └> ""
|  |  |        └> ""
|  |  ├> ""
|  |  |  └> ""
|  |  ├> "AB=="
|  |  |  ├> ""
|  |  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  |     └> ""
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  └> "AA=="
|  |  |     ├> ""
|  |  |     |  └> ""
|  |  |     └> ""
|  |  |        └> ""
|  |  ├> "AA=="
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  └> ""
|  |  |     └> ""
|  |  └> "AAA="
|  |     ├> "AA=="
|  |     |  ├> ""
|  |     |  |  └> ""
|  |     |  └> ""
|  |     |     └> ""
|  |     ├> ""
|  |     |  └> ""
|  |     ├> "AA=="
|  |     |  ├> ""
|  |     |  |  └> ""
|  |     |  └> ""
|  |     |     └> ""
|  |     └> "AA=="
|  |        ├> ""
|  |        |  └> ""
|  |        └> ""
|  |           └> ""
|  ├> "cAB="
|  |  ├> "AB=="
|  |  |  ├> ""
|  |  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  |     └> ""
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  └> "AA=="
|  |  |     ├> ""
|  |  |     |  └> ""
|  |  |     └> ""
|  |  |        └> ""
|  |  ├> "OAB="
|  |  |  ├> "AB=="
|  |  |  |  ├> ""
|  |  |  |  |  ├> ""
|  |  |  |  |  └> ""
|  |  |  |  |     └> ""
|  |  |  |  ├> ""
|  |  |  |  |  └> ""
|  |  |  |  └> "AA=="
|  |  |  |     ├> ""
|  |  |  |     |  └> ""
|  |  |  |     └> ""
|  |  |  |        └> ""
|  |  |  ├> "HAB="
|  |  |  |  ├> "AB=="
|  |  |  |  |  ├> ""
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "AA=="
|  |  |  |  |     └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`base64String (integration) > should be able to shrink '01230123012301230123AB==' 1`] = `
""01230123012301230123AB=="
├> ""
├> "301230123AB="
|  ├> "0123AB=="
|  |  ├> "3AB="
|  |  |  ├> "AB=="
|  |  |  |  ├> ""
|  |  |  |  |  ├> ""
|  |  |  |  |  └> ""
|  |  |  |  |     └> ""
|  |  |  |  ├> ""
|  |  |  |  |  └> ""
|  |  |  |  └> "AA=="
|  |  |  |     ├> ""
|  |  |  |     |  └> ""
|  |  |  |     └> ""
|  |  |  |        └> ""
|  |  |  ├> "AAB="
|  |  |  |  ├> "AB=="
|  |  |  |  |  ├> ""
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "AA=="
|  |  |  |  |     ├> ""
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> ""
|  |  |  |  |        └> ""
|  |  |  |  ├> ""
|  |  |  |  |  └> ""
|  |  |  |  ├> "AB=="
|  |  |  |  |  ├> ""
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "AA=="
|  |  |  |  |     ├> ""
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> ""
|  |  |  |  |        └> ""
|  |  |  |  ├> "AA=="
|  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> ""
|  |  |  |  |     └> ""
|  |  |  |  └> "AAA="
|  |  |  |     ├> "AA=="
|  |  |  |     |  ├> ""
|  |  |  |     |  |  └> ""
|  |  |  |     |  └> ""
|  |  |  |     |     └> ""
|  |  |  |     ├> ""
|  |  |  |     |  └> ""
|  |  |  |     ├> "AA=="
|  |  |  |     |  ├> ""
|  |  |  |     |  |  └> ""
|  |  |  |     |  └> ""
|  |  |  |     |     └> ""
|  |  |  |     └> "AA=="
|  |  |  |        ├> ""
|  |  |  |        |  └> ""
|  |  |  |        └> ""
|  |  |  |           └> ""
|  |  |  ├> "cAB="
|  |  |  |  ├> "AB=="
|  |  |  |  |  ├> ""
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> ""
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "AA=="
|  |  |  |  |     ├> ""
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> ""
|  |  |  |  |        └> ""
|  |  |  |  ├> "OAB="
|  |  |  |  |  ├> "AB=="
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> "AA=="
|  |  |  |  |  |     ├> ""
|  |  |  |  |  |     |  └> ""
|  |  |  |  |  |     └> ""
|  |  |  |  |  |        └> ""
|  |  |  |  |  ├> "HAB="
|  |  |  |  |  |  ├> "AB=="
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`base64String (integration) > should be able to shrink 'ABCD' 1`] = `
""ABCD"
├> ""
├> "CD=="
|  ├> ""
|  |  ├> ""
|  |  ├> ""
|  |  |  └> ""
|  |  └> ""
|  |     ├> ""
|  |     └> ""
|  |        ├> ""
|  |        └> ""
|  |           └> ""
|  ├> "AD=="
|  |  ├> ""
|  |  |  ├> ""
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  └> ""
|  |  |     ├> ""
|  |  |     └> ""
|  |  |        ├> ""
|  |  |        └> ""
|  |  |           └> ""
|  |  ├> ""
|  |  |  └> ""
|  |  ├> "AA=="
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  └> ""
|  |  |     └> ""
|  |  └> "AC=="
|  |     ├> ""
|  |     |  ├> ""
|  |     |  └> ""
|  |     |     ├> ""
|  |     |     └> ""
|  |     |        └> ""
|  |     ├> ""
|  |     |  └> ""
|  |     └> "AB=="
|  |        ├> ""
|  |        |  ├> ""
|  |        |  └> ""
|  |        |     └> ""
|  |        ├> ""
|  |        |  └> ""
|  |        └> "AA=="
|  |           ├> ""
|  |           |  └> ""
|  |           └> ""
|  |              └> ""
|  ├> "BD=="
|  |  ├> ""
|  |  |  ├> ""
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  └> ""
|  |  |     ├> ""
|  |  |     └> ""
|  |  |        ├> ""
|  |  |        └> ""
|  |  |           └> ""
|  |  ├> "AD=="
|  |  |  ├> ""
|  |  |  |  ├> ""
|  |  |  |  ├> ""
|  |  |  |  |  └> ""
|  |  |  |  └> ""
|  |  |  |     ├> ""
|  |  |  |     └> ""
|  |  |  |        ├> ""
|  |  |  |        └> ""
|  |  |  |           └> ""
|  |  |  ├> ""
|  |  |  |  └> ""
|  |  |  ├> "AA=="
|  |  |  |  ├> ""
|  |  |  |  |  └> ""
|  |  |  |  └> ""
|  |  |  |     └> ""
|  |  |  └> "AC=="
|  |  |     ├> ""
|  |  |     |  ├> ""
|  |  |     |  └> ""
|  |  |     |     ├> ""
|  |  |     |     └> ""
|  |  |     |        └> ""
|  |  |     ├> ""
|  |  |     |  └> ""
|  |  |     └> "AB=="
|  |  |        ├> ""
|  |  |        |  ├> ""
|  |  |        |  └> ""
|  |  |        |     └> ""
|  |  |        ├> ""
|  |  |        |  └> ""
|  |  |        └> "AA=="
|  |  |           ├> ""
|  |  |           |  └> ""
|  |  |           └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`base64String (integration) > should be able to shrink 'ABCDABCDABCDABCDABCDABCDABCDABCDABCDA…' 1`] = `
""ABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCD"
├> ""
├> "ABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCD"
|  ├> "CDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCDABCD=="
|  |  ├> "ABCDABCDABCDABCDABCDABCD"
|  |  |  ├> "ABCDABCDABCD"
|  |  |  |  ├> "BCDABCD="
|  |  |  |  |  ├> "ABCD"
|  |  |  |  |  |  ├> "CD=="
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  |           └> ""
|  |  |  |  |  |  |  ├> "AD=="
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  |  |           └> ""
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  ├> "AA=="
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  └> "AC=="
|  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |     |  ├> ""
|  |  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |  |     |     ├> ""
|  |  |  |  |  |  |  |     |     └> ""
|  |  |  |  |  |  |  |     |        └> ""
|  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |  |     └> "AB=="
|  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |        |  ├> ""
|  |  |  |  |  |  |  |        |  └> ""
|  |  |  |  |  |  |  |        |     └> ""
|  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |        |  └> ""
|  |  |  |  |  |  |  |        └> "AA=="
|  |  |  |  |  |  |  |           ├> ""
|  |  |  |  |  |  |  |           |  └> ""
|  |  |  |  |  |  |  |           └> ""
|  |  |  |  |  |  |  |              └> ""
|  |  |  |  |  |  |  ├> "BD=="
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  |  |           └> ""
|  |  |  |  |  |  |  |  ├> "AD=="
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  |  |  |           └> ""
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  ├> "AA=="
|  |  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  |  └> "AC=="
|  |  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |  |     |  ├> ""
|  |  |  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |  |  |     |     ├> ""
|  |  |  |  |  |  |  |  |     |     └> ""
|  |  |  |  |  |  |  |  |     |        └> ""
|  |  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |  |  |     └> "AB=="
|  |  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |  |        |  ├> ""
|  |  |  |  |  |  |  |  |        |  └> ""
|  |  |  |  |  |  |  |  |        |     └> …
|  |  |  |  |  |  |  |  |        └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
